hips/hip-9999.md

```diff
@@ -0,0 +1,81 @@
+
+The second part of this proposal would include the tarball digest of a chart in the metadata. This aspect is less fleshed out in code yet, but would be a nice alternative for installing a local chart from a tarball when the repository URL would not be known. The reason this is useful is because it can be used when interacting with other third-party databases, such as artifact-hub.
+
+## Motivation
+
+The primary motivation is to have a helm-native way to track where a chart came from. Many of the motivations for this feature can be found in this issue: https://github.com/helm/helm/issues/4256. To note a couple:
+- An admin would like to know where a chart came from only by querying the cluster itself. Currently the only way to know where a chart came from is to keep track of it externally.
+- Third-party tooling that provides recommendations on helm upgrades can query the cluster to get the repository URL to aid in this recommendation. See [nova](https://github.com/FairwindsOps/nova) as an example.
+- This metadata field could also populate output for the `helm ls` command in the future. That may be outside the scope of this particular feature.
+
+## Rationale
+
+In order to query for this in the cluster, the repository URL will need to be somewhere in the release object. The chart metadata portion of the release object already contains similar data and makes the most sense for this purpose.
+
+## Specification
+
+Nothing would change from the user's perspective. With some modifications in the current code it's possible to add this metadata into the release object transparently.
+
+The current idea for this implementation is to add a field in both the `ChartDownloader` and `Metadata` (in the `pkg/chart/metadata.go` file) structs to hold the repository URL. The Downloader is the logical place to derive the proper URL and then populate the Metadata field.
+
+More specifically, the URL is included in the local repository index file for a locally named repository. For instance, running `helm repo add [NAME] [URL]` would add the repository to the index file with the `[URL]` passed as the real location to download charts. In the helm code, the downloader reads this file to find the URL which is what gets bubbled up to the release object.
+
+If for any reason this data can not be gathered at install/upgrade time, then that particular field would be omitted from the release object metadata.
```

Comment:
I do not think omitting this field from the release metadata fits with the use cases in the motivation section above. Omitting the field entirely – as is proposed now – would not allow users to differentiate between charts installed from a local path vs charts installed from an earlier version of Helm that did not include this feature.

Alternatively Helm could store an indicator that the chart was installed from a local path (to a packed or unpacked chart) if it was installed using methods 2 or 3 from `helm install --help`:

```
There are five different ways you can express the chart you want to install:

1. By chart reference: helm install mymaria example/mariadb
2. By path to a packaged chart: helm install mynginx ./nginx-1.2.3.tgz
3. By path to an unpacked chart directory: helm install mynginx ./nginx
4. By absolute URL: helm install mynginx https://example.com/charts/nginx-1.2.3.tgz
5. By chart reference and repo url: helm install --repo https://example.com/charts/ mynginx nginx
```

Since only in the case of methods 2 and 3 there is no accessible URL, it could be a value convention such as: `"repoURL": "path"`.

Reply:
Yeah, I see your point. I could go either way on this one.

Reply:
Love this idea, and I updated the HIP to include it.